### PR TITLE
Fix multiprocess download hanging in notebooks

### PR DESCRIPTION
**Pull request by dbczumar**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/file_utils.py [hunk 1]
```diff
@@ -643,25 +645,50 @@ def parallelized_download_file_using_http_uri(
...
-                http_uri=http_uri,
-            )
-            futures[i] = fut
+    download_procs = {}
+
+    def await_active_procs(max_active_procs, synchronous):
+        while len(download_procs) > max_active_procs:
+            completed_downloads_idxs = []
+            for idx, download_proc in download_procs.items():
+                return_code = download_proc.wait() if synchronous else download_proc.poll()
```
Comment (apurva-koti) [inline mlflow/utils/file_utils.py:654]: Wouldn't this wait for each download_proc in order?

**Review by apurva-koti — COMMENTED** [review]
Do we see a significant change in the speedup factor after doing this? + have we done checksum comparisons?

File: mlflow/utils/file_utils.py [hunk 2]
```diff
@@ -643,25 +645,50 @@ def parallelized_download_file_using_http_uri(
...
+                    completed_downloads_idxs.append(idx)
+                    if return_code != 0:
+                        stdout, _ = download_proc.communicate()
+                        failed_downloads[i] = json.loads(stdout)
+
+            for completed_idx in completed_downloads_idxs:
+                del download_procs[completed_idx]
+
+    for i in range(starting_index, num_requests):
+        await_active_procs(max_active_procs=MAX_PARALLEL_DOWNLOAD_WORKERS, synchronous=False)
```
Comment (apurva-koti) [inline mlflow/utils/file_utils.py:665]: Should we await once before launching all the commands?

**Comment by dbczumar** [comment]
> Do we see a significant change in the speedup factor after doing this? + have we done checksum comparisons?

It's still quite fast after some additional changes: https://github.com/mlflow/mlflow/pull/8352. This needs serious testing though, including checksums, ensuring nothing hangs, retry logic for 401s / 403s (this was broken before).

I'm closing this in favor of https://github.com/mlflow/mlflow/pull/8352.

**Pull request closed without merging**